### ISSUE TITLE
Possibly bad code generated for const inside classes

### ISSUE THREAD

**Issue by teobugslayer** [Microsoft/TypeScript#3068]
Hi all,
I think I found an issue with the new TSC 1.5 const support.

Consider this code:

````TypeScript
const A = "5"

class P {
	const K = "6"
	const J = "7"
}
````

In Playground, the generated code is:

```JavaScript
var A = "5";
var P = (function () {
    function P() {
        this.const = K = "6";
        this.const = J = "7";
    }
    return P;
})();
```

which I believe is broken

**Comment by vladima** [comment]
This code is syntactically invalid and for this kind of errors we never make any promises about quality of emit
```
node built\local\tsc.js ..\a.ts
../a.ts(4,11): error TS1005: '=' expected.
../a.ts(5,11): error TS1005: '=' expected.
```

**Comment by DanielRosenwasser** [comment]
To clarify what @vladima just said, `const` is an invalid modifier for a class member - you can use an accessor if you want it to be read-only.

Though it's kind of odd that we don't have a better error for this.

**Comment by teobugslayer** [comment]
Hello @vladima , @DanielRosenwasser,

Please reconsider closing this issue. 

I believe that silent bad code generation is a wrong behavior. If possible, change the compiler to check for this condition and emit an error.

**Comment by mhegazy** [comment]
@teobugslayer the compiler indeed does emit an error.. see the red squiggles in the playground:
![capture](https://cloud.githubusercontent.com/assets/8000722/7532442/c4cf5588-f51b-11e4-89e4-62d583fc8cc5.PNG)

**Comment by teobugslayer** [comment]
Hello @mhegazy,

The error given is

```
error TS1005: '=' expected.
```

which is not very helpful.

**Comment by mhegazy** [comment]
That is fair. We would take a PR to give a better error message.

**Comment by CyrusNajmabadi** [comment]
If you do not want code generation on errors, use the –noEmitOnError flag.

Cheers!

           -- Cyrus

**Comment by RyanCavanaugh** [comment]
> I believe that silent bad code generation is a wrong behavior.

This is already the behavior by default. If you have syntactic errors in your file, tsc will not emit any code.

The website overrides this behavior and updates the right-hand side anyway, but this doesn't affect how the commandline tools work.

**Comment by CyrusNajmabadi** [comment]
That's not the case anymore.  tsc will emit in the presence of syntax errors.  Indeed, we'll basically always emit.  The only cases were we don't are things like:

1. There was a tsconfig file, but we couldn't parse it.
2. There were irreconcilable command line errors.

Once we've decided that your options are ok, and we know what files to compile, then we'll pretty much always emit (unless you have noEmitOnErrors).

**Comment by RyanCavanaugh** [comment]
When did we change that? :frowning:

**Comment by CyrusNajmabadi** [comment]
We don't have a good error here because:

1. 'const' isn't a modifier in the traditional sense.  After all, you can just write "const a", in which case it's a symbol keyword (like 'class', 'var', etc.)
2. 'const' is a legal class property name.   

So we parse out 'const' as a legal name, and then need something valid to follow it.  We don't see anything, so we report the simplest thing that would make the code compilable.  In this case, adding an '=' sign after 'const' *would* make it compilable.

The appropriate fix here is to not absorb a keyword as a class property name if it is followed by an identifier/keyword on the same line.  Then we won't consume the 'const' as a property name.  We'll then want to allow consuming 'const' along with other modifiers as a class member modifier.  We'll then have to report that 'const' isn't legal on a class member during our grammar modifier checks.

**Comment by CyrusNajmabadi** [comment]
We changed it during 1.5.  It became much simpler and more consistent.  Users don't have to know which set of compiler errors woudl block compilation and which would not.  For example, a syntactic block would mean that the following would block compilation:

```function foo(a?, b)```

However, this would not:

```function foo(public a)```

(The reason being that the first is actually disallowed syntactically, but the second is not)

It also added a lot of complexity as we had to put all sorts of errors into our syntactic pass.  Now the syntactic pass is super simple.  And the semantic pass simply checks for pretty much all the interesting errors in our language (outside of declaration-specific errors).

**Comment by danquirk** [comment]
#1915 covers reverting this behavior (emit despite errors in VS).

**Comment by teobugslayer** [comment]
Thanks all who enlighten me on the issue, it helped me much understanding the concepts involved!
